src/core/geometry/qgscurve.cpp

```diff
@@ -434,3 +434,27 @@ bool QgsCurve::snapToGridPrivate( double hSpacing, double vSpacing, double dSpac
 
   return outSize >= 4 || ( !isClosed() && outSize >= 2 );
 }
+
+double QgsCurve::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for curves (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  if ( fromVertexNumber < 0 || fromVertexNumber >= numPoints() ||
+       toVertexNumber < 0 || toVertexNumber >= numPoints() )
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Default implementation: calculate distance along segmentized curve
+  std::unique_ptr< QgsLineString > lineString( curveToLine() );
+  if ( !lineString )
+    return -1.0;
+
+  return lineString->distanceBetweenVertices( fromVertex, toVertex );
+}
```

Comment:
This isn't needed, is it? All the QgsCurve subclasses are implementing the method. I'd make it pure virtual instead. (Which is acceptable to do and doesn't count as an api break, as no-one should be subclassing QgsCurve from python).

Reply:
Indeed, it's a remnant from my first version. It's been removed and it's a virtual method.